storage/include/sparse_merkle/base_types.h

```diff
@@ -110,6 +145,25 @@ class NibblePath {
     num_nibbles_ += 1;
   }
 
+  // Pop the last nibble off the path
+  Nibble pop_back() {
```

Comment:
Just curious, what all this Nibble stuff is used for in merkle trees?

Reply:
In order to reduce the number of database accesses we batch internal nodes of the sparse merkle tree into a 4 level tree. A four level tree has 2^4 = 16 possible leaves, each uniquely identifiable by a nibble (4 bits).  Some of these leaves are pointers to other internal nodes. We can therefore walk from the root of the tree to the actual leaf being searched by taking the hash key, and pulling off a nibble at a time to retrieve the "next" internal node from the database. As we take each nibble from the beginning of the hash of the real key, we build up a `NibblePath` that (along with versions - ignore for now) uniquely identifies a node in the tree. We can walk back up the tree, by popping off the tail nibble off the path to get to the prior internal node.

Picture our tree looks like the following: 3 internal nodes with the leaf pointer precisely in the leaf of the 3rd internal node. Remember from prior discussions we don't need to store any deeper in the tree if there are no subtrees underneath. Root is just an internal node with an empty nibble path.

Root -> 3 -> F -> A (Leaf Pointer)

We hash a key = 3FA12B9...
We need to insert it by walking the tree. We get the first internal node after the root = 3, get the second at 3F, and see that 3FA, a leaf in 3F internal node is empty. We then decide to store the pointer to the leaf the there.